[웹개발 3주차 미션] 구현 완료했습니다. 리뷰 부탁드립니다.

java-saeng opened the pull request:
# 공부할 점
- ModelAndView와 Model 차이점은 뭘까?
- GeneratedValue의 strategy에 대해 깊이 파볼 것

# 느낀 점
- QuestionController에서 질문 목록 외 나머지 URI에는 /questions이 들어있었는데 RequestMapping으로 전부 통일하고 싶은 마음이 들었습니다.
- 왜 명세에서는 Optinal을 반환할 때 .get()을 사용할까?
- 명세에서 CrudRepository를 extends 하라는 것을 보고, 미션이 정말 순차적으로 잘 짜여져있다는 생각했습니다.

# 궁금한 점
- "객체지향의 사실과 오해" pg 56에서 "메시지를 해석하고 그에 반응해서 상태를 변경할 지 여부는 전적으로 메시지 수신자의 자율적인 판단에 따른다." 라는 부분을 읽고 TDA 법칙이 생각났습니다.
- User를 update할 때, controller에서 password가 일치하는지 판단하는 "행동"은 객체가 판단하는게 더욱 객체지향적일까 라는 생각을 했습니다.

java-saeng committed:
refactor : 패키지 구조 변경

domain과 controller 별로 나누지 않고 UserController와 User는 user 패키지에 저장

java-saeng committed:
feat : 질문하기 form 맵핑

java-saeng committed:
feat : 질문하기, 질문목록, 질문 상세보기 기능 구현

질문목록을 보기 위한 list.html은 user의 list.html을 참고함

상세보기 기능 구현 시 index가 1부터 시작하여 @index_1 사용

java-saeng committed:
feat : css와 js의 경로를 상대 경로에서 절대 경로로 변경

기존 base에 있던 css와 js는 상대 경로 설정
updateForm.html을 불러올 때 경로가 달라져서 상대 경로를 절대 경로로 변경

java-saeng committed:
feat : 개인정보수정 화면 구현

list.html에 있는 수정 버튼의 링크 URI 변경
updateForm.html은 수정 페이지로, 변경할 수 있는 기존 정보 들어있음

java-saeng committed:
refactor : PathVariable에서 index 명시

java-saeng committed:
feat : User 정보 수정하는 기능 구현 및 id를 통해 회원 조회 리팩토링

User 클래스에 수정, 비밀번호 검사 메서드 생성
UserController에서 id를 통해 회원 조회하는 로직 리팩토링 및 회원 정보 수정 기능 구현

saint6839 committed:
refactor : properties -> yml 변경

saint6839 committed:
refactor : index를 id로 변경

saint6839 committed:
feat : id 필드, update 메서드, get 메서드 추가

saint6839 committed:
feat : id 필드, get 메서드 추가

saint6839 committed:
refactor : controller DB 연동

- CRUD Repository 연동
- create() : 회원가입 메서드
- list() : 유저 목록 조회
- show() : 유저 정보 조회
- updateForm() : 수정 폼 보여줌
- update() : 사용자 정보 수정

saint6839 committed:
feat : CrudRepository 확장

saint6839 committed:
feat : QuestionController DB 연동

-create() : 질문 생성
-list() : 질문 목록 조회
-qnaInfo() : 질문 조회
-update() : 질문 수정
-updateForm() : 질문 수정 폼 불러오기
-remove() : 질문 삭제

saint6839 committed:
refactor : qna -> question, 삭제 버튼 경로 변경

saint6839 committed:
feat : 질문 수정 폼 추가

java-saeng committed:
refactor : 회원 조회 기능 메서드 extract

아이디를 통해 회원 조회 기능 중복되어 메서드 extract
필요없는 List 삭제

java-saeng committed:
refactor : 질문 조회 기능 메서드 extract

아이디를 통해 질문 조회 기능 중복되어 메서드 extract

java-saeng committed:
refactor : user update 시 TDA 법칙 적용

hyukjin-lee commented on src/main/java/codesquad/qua/QuestionController.java:
```diff
@@ -0,0 +1,61 @@
+@Controller
+public class QuestionController {
+
+    @Autowired
```
의존성 주입(DI) 란 무엇이고 스프링에서 제공하는 방법에는 Autowired 말고 어떤 대안들이 있을까요?  
+ Autowired 방식의 단점이 무엇일까요?

java-saeng replied:
"스프링에서 제공하는 방법에는 Autowired 말고 어떤 대안들이 있을까요" 에 대한 궁금한 점이 있습니다. 

원래 다른 대안들은 생성자 주입, setter 주입이라고 생각했는데 이 주입 방법들 또한 Autowired를 사용한 것으로 다른 답이 있구나라는 생각이 들어 검색을 해봤습니다.

제 검색 능력이 부족한 탓인지, Autowired를 제외한 방법에는 @Inject, @Resource를 알게 됐습니다. 하지만 해당 어노테이션들은 패키지가 javax로 되어있어서 스프링에서 제공하는 방법이 아니라고 생각이 드는데 혹시 키워드를 알 수 있을까요??


제가 구글링해본 문장입니다.
what is method of dependency injection in spring

spring provided dependency injection system instead of autowired

hyukjin-lee replied:
그러면 autowired 명시를 통한 필드 주입의 단점과 그에 대한 대안으로 질문을 좀 더 구체화해서 여쭤볼게요 ~

hyukjin-lee commented on src/main/java/codesquad/qua/QuestionController.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    @GetMapping("/questions/{id}")
+    public String qnaInfo(Model model, @PathVariable("id") Long id) {
```
PathVariable 에 long 으로 사용해도 될까요?

java-saeng replied:
저는 primitive를 지양해야 한다고 생각합니다.
primitive type은 데이터베이스에 값이 없더라도 기본값을 유지합니다.
따라서 long을 사용하기보다는 reference type인 Long을 사용하는 것을 지향합니다. 

hibernate에서는 persistence class의 식별자 값을 non-primitive 사용하는 것이 좋다고 합니다!!

https://docs.jboss.org/hibernate/orm/3.3/reference/en/html/persistent-classes.html#persistent-classes-pojo-identifier

hyukjin-lee replied:
여기는 hibernate entity 가 아닌데 java 전역에서 primitive type 을 지양해야한다고 말씀하시는걸까요~?

java-saeng replied:
Question의 id 값에 대해 long으로 사용해도 되는가에 대해 말씀하신줄 알았습니다.

java-saeng replied:
처음에 Question의 id에 long을 사용해도 되는지에 대해 말씀하신걸로 이해했습니다.

long과 Long의 차이는 null 의 차이라고 생각합니다.
아래의 코드와 같이 {id}값이 들어오지 않게 되면 아래와 같이 런타임 에러가 발생합니다.

하지만 식별자 값으로 생각하지 않고 단순히 Long 대신에 long을 사용할 수 있느냐에 대한 저의 생각은 사용할 수 있다고 생각합니다. id가 String형으로 들어와도 long이나, Long 둘 다 translate 해주기 때문입니다.

![image](https://user-images.githubusercontent.com/62413589/160580385-f2a5fdfd-b7c9-4124-87b0-9ccd2126906a.png)

hyukjin-lee commented on src/main/java/codesquad/qua/QuestionController.java:
```diff
@@ -0,0 +1,61 @@
+    public String update(Question question, @PathVariable("id") Long id) {
+        Question findQuestion = getQuestionById(id);
+
+        findQuestion.update(question);
```
findQuestion 은 메서드명 같은데 다른 변수명을 사용해보면 어떨까요?

hyukjin-lee commented on src/main/java/codesquad/qua/QuestionController.java:
```diff
@@ -0,0 +1,61 @@
+
+    @GetMapping("/questions/{id}/updateForm")
+    public String updateForm(Model model, @PathVariable("id") Long id) {
+        Question findQuestion = getQuestionById(id);
```
get 과 find 라는 단어는 각각 어떤 경우에 사용할까요?

java-saeng replied:
get은 해당 객체가 고유하게 소유한 정보에 대해 얻고 싶을 때 사용하고,

find는 정보를 얻기 위해 다른 객체를 사용하여 정보를 찾을 때 사용할 것 같습니다.

따라서 getQuestionById는 get보다는 find가 좋을 것 같습니다.

hyukjin-lee replied:
이 글을 추천드립니다
https://tuhrig.de/find-vs-get/

java-saeng replied:
글을 읽어보고 나서 get과 find의 차이를 JpaRepository와 CrudRepository에 연관지어 생각해봤습니다.

find는 해당 항목이 존재하지 않을 수 있다는 가정 하에, caller가 예외처리를 해줘야하기 때문에 Optional로 반환하고,

get은 일치 항목이 존재한다고 가정합니다. 따라서 존재하지 않게 되면 JpaRepository 자체 내에서 에외를 발생시킵니다.

그렇다면 getQuestionById는 모두 Question이 존재한다는 가정을 하기 때문에 get이 적절해보입니다.

올려주신 글에서 궁금한 점이 있습니다.

get을 사용해서 없을 경우에 예외를 던지는데, 이때 "caller doesn't need to handle the inconsistency of the persistenc layer" 라는 말이 getById(=getOne)은 getReference를 호출, 즉, 프록시 객체를 조회하기 때문에 데이터의 일관성을 신경쓰지 않아도 된다는 말일까요??

hyukjin-lee commented on src/main/java/codesquad/qua/QuestionController.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    private Question getQuestionById(Long id) {
+        return questionRepository.findById(id).get();
```
.get() 을 바로 사용하는걸 지양해야하는 이유가 무엇일까요?

java-saeng replied:
Optional 의 주된 목적은 메서드가 반환할 결과값이 "없음"을 명백하게 표현하고, NPE를 방지하기 위해 추가된 기능입니다. 하지만 get을 바로 사용하게 되면 Optional 을 사용하는 목적에 부합하지 않고, NPE가 발생할 수 있기 때문에 지양하는 것 같습니다.

hyukjin-lee commented on src/main/java/codesquad/qua/QuestionRepository.java:
```diff
@@ -0,0 +1,6 @@
+
+import org.springframework.data.repository.CrudRepository;
+
+public interface QuestionRepository extends CrudRepository<Question, Long> {
```
JpaRepository 와 CrudRepository 의 차이점이 무엇일까요?

java-saeng replied:
```
public interface JpaRepository<T, ID> extends PagingAndSortingRepository<T, ID>, QueryByExampleExecutor<T> {

public interface PagingAndSortingRepository<T, ID> extends CrudRepository<T, ID>
```

JpaRepository의 패키지 구조 `org.springframework.data.jpa.repository`

CrudRepository의 패키지 구조는`org.sporg.springframework.data.repository` 입니다.

즉, JpaRepository는 JPA에 특화된 기능들이 있습니다. (영속성 컨텍스트에 따른 1차 캐시, 더티 체킹, 지연 로딩, 쓰기 지연 등)

CrudRepository는 jpa, mongo-DB, redis 등 CRUD는 모두 비슷하므로 공통 인터페이스가 제공되는 것입니다. 이러한 공통적인 인터페이스들은 spring-data-commons에 들어있습니다.

![image](https://user-images.githubusercontent.com/62413589/160956957-a3cbf8b7-9a17-4842-9f97-1aebbd82a562.png)

hyukjin-lee commented on src/main/java/codesquad/user/User.java:
```diff
@@ -14,31 +23,46 @@ public String getUserId() {
 
-    public void setEmail(String email) {
-        this.email = email;
+    private boolean validatePassword(User user) {
```
equalsPassword(string password) 는 어떤가요?

java-saeng replied:
비밀번호를 인증한다는 생각에 validate를 사용했던 것 같습니다.
자세히 알아보니 "인증"이라는 뜻은 사용자가 "누구"인지 확인하는 절차로 현재 메서드의 기능은 단지 비밀번호가 같은지 확인하는 기능으로써  equalsPassword가 적합한 것 같습니다. valiedate는 로그인할 때 사용자인지 확인하는 메서드에서 사용하면 좋을 것 같습니다.

parameter로 User보다는 String을 주는게 맞는 것 같습니다. 위와 마찬가지로 해당 메서드는 비밀번호가 맞는지 확인하는 행동을 하기 때문에 User보다는 password를 주는게 좋을 것 같습니다.

hyukjin-lee commented on src/main/resources/application.yml:
```diff
@@ -7,10 +7,10 @@ spring:
   jpa:
     hibernate:
-      ddl-auto: create
+      ddl-auto: create-drop
```
이 옵션의 의미와 어떤 다른 선택지가 있는지 설명가능하실까요~?

java-saeng replied:
제가 알고 있는 ddl-auto option은 create, create-drop, update, validate, none이 있습니다. ddl-auto의 default 값은 create-drop으로 알고 있고, ddl-auto를 적용하지 않기 위해서는 none을 사용해야한다고 알고있습니다.
![image](https://user-images.githubusercontent.com/62413589/160772415-163d6325-0fbc-4d50-9cf5-3ff15e763247.png)


create
- 기존 테이블을 삭제하고 새로운 스키마를 만듭니다.

create-drop
- create와 마찬가지로 기존 테이블 삭제하고 새로운 테이블 생성
- 해당 세션이 끝나면 DB를 drop 한다
- 유닛 테스트에서 용이하게 사용

update
- 기존 스키마에서 변경된 스키마에만 적용한다

validate
- 테이블과 컬럼이 존재하는지 여부 확인. 없으면 예외 발생

none
- DDL 옵션을 사용하지 않음

hyukjin-lee reviewed (approved):
고생하셨습니다~

java-saeng committed:
chore : validatePassword 메서드 이름 및 매개변수 변경

validatePassword -> equalsPassword
- validate는 인증이라는 뜻으로 equals가 더 나아보임

매개변수 user -> password로 변경
- 비밀번호가 맞는지 확인하는 기능인데, 굳이 user 엔티티를 전달하는 것보다는 필요한 데이터인 비밀번호만 주는 것이 낫다고 생각함

java-saeng committed:
chore : findQuestion -> savedQuestion 변경

findQuestion은 메서드 이름 같으므로 savedQuestion으로 변경

java-saeng committed:
fix : Optional 예외 처리

get을 사용하지 않고 명시적으로 NULL 처리

java-saeng committed:
chore : foundUser -> savedUser
